### PR TITLE
add support for bigquery external connections.

### PR DESCRIPTION
**Pull request by 600lyy**
### Change description

Add support for Bigquery connection that can be used to create external tables (including biglake) through KCC Bigquery.
Expose the serviceaccount autogenerated  in the observed state to be referenced in IamPolicies

issue number: https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/1392
b/326180149

### Tests you have done

> go test -v -tags=integration ./pkg/controller/dynamic/ -test.run TestCreateNoChangeUpdateDelete -run-tests cloudresourceconnection -timeout 900s

> go test -v -tags=integration ./config/tests/samples/create -test.run TestAll -run-tests bigqueryconnectionconnection

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/resourceskeleton/testdata/uri-skeleton.yaml
Comment (maqiuyujoyce) [inline pkg/resourceskeleton/testdata/uri-skeleton.yaml:795]: `v1alpha1`. This is not a v1beta1 resource.

File: pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go:20]: Is this resource not updatable at all?

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml
Comment (maqiuyujoyce) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml:311]: All the comments in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1682/files#diff-508baa3dabf00328c18fcd828339759a13777f9c08481225ebffd438d071070f is still applicable to the CRD.

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (600lyy) [reply config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:18]: Resolved

File: pkg/snippet/snippetgeneration/snippetgeneration.go
Comment (600lyy) [reply pkg/snippet/snippetgeneration/snippetgeneration.go:39]: ok, resolved.

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml
Comment (600lyy) [reply pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml:18]: sounds good!

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml
Comment (600lyy) [reply pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml:1]: sounds good!

File: pkg/resourceskeleton/testdata/uri-skeleton.yaml
Comment (600lyy) [reply pkg/resourceskeleton/testdata/uri-skeleton.yaml:795]: thanks!

File: pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go
Comment (600lyy) [reply pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go:1]: thanks!

File: pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go
Comment (600lyy) [reply pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go:20]: For connection type "cloudResource", yes update operation is not supported.

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml
Comment (600lyy) [reply pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1beta1/cloudresourceconnection/create.yaml:1]: I think the directory name should be #bigqueryconnectionconnection/v1alpha1/cloudresourceconnection#?
Bigquery connection supports different types of external connections, and cloudresource is just one of those

File: pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go
Comment (600lyy) [reply pkg/test/resourcefixture/contexts/bigqueryconnectionconnection_context.go:18]: Thanks!

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (maqiuyujoyce) [reply config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:25]: This is a tricky interface because `serviceAccountId` is actually output-only but got added under `spec`. Though it's okay to be used as a workaround in v1alpha1.

In order to populate its value under `status.observedState`, you should adopt the workaround in #2032.

File: pkg/test/resourcefixture/contexts/bigqueryconnection_context.go
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/contexts/bigqueryconnection_context.go:20]: Could you add a comment explain why `SkipUpdate` is `true`?

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (maqiuyujoyce) [inline config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:1]: The filename should be `bigqueryconnection_v1alpha1_bigqueryconnectionconnection.yaml`: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.Samples.md#file-naming

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml
Comment (maqiuyujoyce) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml:311]: Could you either turn `database`, `instanceId`, etc into reference fields accordingly, or maybe just ignore `cloudSql` and `cloudSpanner` fields completely in the service mapping?

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (600lyy) [reply config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:25]: @maqiuyujoyce Thanks!
Do you want me to include the workaround in this PR, or will it be a separate one?

File: pkg/test/resourcefixture/contexts/bigqueryconnection_context.go
Comment (600lyy) [reply pkg/test/resourcefixture/contexts/bigqueryconnection_context.go:20]: Sounds good!

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (maqiuyujoyce) [reply config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:25]: Yes, you'll need to include the workaround in https://github.com/GoogleCloudPlatform/k8s-config-connector/commit/ee1d4b63427d69c57c7d3dbe8f4b2ed300a706fb to populate `status.observedState.serviceAccountId` properly during the reconciliation.

In addition, I suggest you run the VCR test and golden object comparison using the command to ensure this field is always populated properly:

```
E2E_KUBE_TARGET=envtest RUN_E2E=1 \
E2E_GCP_TARGET=vcr VCR_MODE=replay GOLDEN_OBJECT_CHECKS=1 \
go test -timeout 3600s -v ./tests/e2e -run TestAllInSeries/fixtures/^bigqueryconnectionconnection$
```

You'll want to generate/set up VCR cassette and the golden object first.
* Instructions for VCR testing: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/tests/README.VCRTesting.md
* To generate the golden object, you should add the environment variable `WRITE_GOLDEN_OUTPUT=1` during the test run. Some potentially helpful doc can be found here: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/mockgcp/README.md#capture-http-golden-logs

@gemmahou should be a good contact if you have any questions about the test.

**Comment by yuwenma** [comment]
/hold

Can you rebase this PR to the HEAD and make sure it pass the `scenarios-tests` and `tests-e2e-fixtures`? We added the golden check as a PRESUBMIT requirements.  You may need to run `hack/compare-mock fixtures/<the new test you add in this PR>` to generate the golden log.

File: config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml
Comment (maqiuyujoyce) [reply config/samples/resources/bigqueryconnectionconnection/bigqueryconnectionconnection_v1alpha1_bigqueryconnectionconnection.yaml:25]: Here is the doc about establishing the test for observed state: https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2128/files#diff-4ba365324b277bbbedf753f1d49fe7b02c3ee05cbccca1c72bd658f30ca14da8

Example can be found in the same PR.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml
Comment (600lyy) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com.yaml:311]: I will make them ignored fields for now. 
Support for cloudSql and cloudSpanner can be requested separately, https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/2149
But I don't want to hold up the current alpha release for that.

File: pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go
Comment (yuwenma) [inline pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go:128]: Is this a Google service account? If so, can you name it `gcpServiceAccount` to distinguish with the Kubernetes service account (which in CRD is normally called `ServiceAccount`)?

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/_generated_object_bigqueryconnectionconnection.golden.yaml
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/_generated_object_bigqueryconnectionconnection.golden.yaml:7]: Can you remove the state-into-spec annotation? We are deprecating this annotation completely?

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve
/hold I'd like to have @gemmahou take another review, focusing on VCR and the TF/DCL general requirements

File: pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/_generated_object_bigqueryconnectionconnection.golden.yaml
Comment (600lyy) [reply pkg/test/resourcefixture/testdata/basic/bigqueryconnection/v1alpha1/bigqueryconnectionconnection/_generated_object_bigqueryconnectionconnection.golden.yaml:7]: sounds good, thanks!

File: pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go
Comment (600lyy) [reply pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go:131]: @yuwenma  
They're auto-generated, wouldn't be manual changes reverted if the file is regenerated?

File: pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go
Comment (600lyy) [reply pkg/clients/generated/apis/bigqueryconnection/v1alpha1/bigqueryconnectionconnection_types.go:128]: Same as above

**Comment by gemmahou** [comment]
Unit test failed:
```
=== Failed
=== FAIL: tests/apichecks TestCRDsAcronyms (1.76s)
    utils.go:99: unexpected diff in testdata/exceptions/acronyms.txt:   (
          	"""
          	... // 43 identical lines
          	[acronyms] crd=bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com version=v1alpha1: field ".spec.azure.objectId" should be ".spec.azure.objectID"
          	[acronyms] crd=bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com version=v1alpha1: field ".spec.cloudResource.serviceAccountId" should be ".spec.cloudResource.serviceAccountID"
        - 	[acronyms] crd=bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com version=v1alpha1: field ".spec.cloudSql.instanceId" should be ".spec.cloudSql.instanceID"
        - 	[acronyms] crd=bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com version=v1alpha1: field ".spec.cloudSql.serviceAccountId" should be ".spec.cloudSql.serviceAccountID"
        + 	[acronyms] crd=bigqueryconnectionconnections.bigqueryconnection.cnrm.cloud.google.com version=v1alpha1: field ".status.observedState.cloudResource.serviceAccountId" should be ".status.observedState.cloudResource.serviceAccountID"
          	[acronyms] crd=bigquerydatasetaccesses.bigquery.cnrm.cloud.google.com version=v1alpha1: field ".spec.dataset.dataset.datasetId" should be ".spec.dataset.dataset.datasetID"
          	[acronyms] crd=bigquerydatasetaccesses.bigquery.cnrm.cloud.google.com version=v1alpha1: field ".spec.dataset.dataset.projectId" should be ".spec.dataset.dataset.projectID"
          	... // 465 identical lines
          	"""
          )
```
The `Id` in CRD and other places(e.g. test yaml) needs to be changed to `ID`

**Review by regadas — APPROVED** [review]

**Comment by yuwenma** [comment]
/lgtm

Thank you! I'll defer to @gemmahou to unhold the bit

**Pull request merged**
